text/0000-dynamic-stylesheet-replacement.md

```diff
@@ -0,0 +1,260 @@
+export const StyleMixin = (superclass, stylesheets) => {
+    return class extends superclass {
+        static stylesheets = stylesheets;
+    };
+};
+```
+
+```js
+// dark.js
+import Base from './base.js';
+import { StyleMixin } from './styleMixin.js';
+import darkStylesheets from './dark.css';
+
+export default StyleMixin(Base, darkStylesheets);
+```
+
+```js
+// light.js
+import Base from './base.js';
+import { StyleMixin } from './styleMixin.js';
+import lightStylesheets from './light.css';
+
+export default StyleMixin(Base, lightStylesheets);
+```
+
+```html
+<!-- wrapper.html -->
+<template lwc:if={darkMode}>
+    <x-dark></x-dark>
+</template>
+<template lwc:else>
+    <x-light></x-light>
+</template>
+```
+
+```js
+// wrapper.js
+export default class extends LightningElement {
+  @api darkMode;
+}
+```
+
+</details>
+
+This is a lot of code! Plus it requires the additional wrapper component.
+
+Another goal of this RFC is to fully remove the [hacks that component authors are using](https://github.com/salesforce/lwc/issues/2826) to apply different stylesheets to the same template:
+
+```js
+import template from './template.html';
+import darkStylesheets from './dark.css';
+import lightStylesheets from './dark.css';
+
+export default class extends LightningElement {
+  @api darkMode;
+  
+  render() {
+    // DEPRECATED - DO NOT DO THIS!
+    template.stylesheets = darkMode ? darkStylesheets : lightStylesheets;
+    template.stylesheetToken = darkMode ? 'dark' : 'light';
+    return template;
+  }
+}
+```
+
+Programmatic stylesheets can replace some of these hacks, but not all. This proposal aims to replace _all_ of these use cases.
+
+## Detailed design
+
+LWC components are already able to [render multiple templates](https://lwc.dev/guide/html_templates#render-multiple-templates). However, the assumption is that component authors will
+only switch between different HTML files:
+
+```js
+import a from './a.html';
+import b from './b.html';
+
+export default class extends LightningElement {
+  @api shouldUseA;
+  
+  render() {
+    return shouldUseA ? a : b;
+  }
+}
+```
+
+(These HTML files may have their own implicitly-associated stylesheets, but component authors cannot reuse the same HTML file with different stylesheets.)
+
+This proposal augments the current behavior of the `render()` function. In addition to being able to return a template, the `render()` function can also return an object containing the `template` and `stylesheets` properties:
+
+```js
+import template from './template.html';
+import a from './a.css';
+import b from './b.css';
+
+export default class extends LightningElement {
+  @api shouldUseA;
+  
+  render() {
+    return {
+      template,
+      stylesheets: shouldUseA ? a : b
+    }
+  }
+}
+```
+
+The stylesheets are injected in the following order:
+
+1. Stylesheets implicitly associated with the `template`
+2. Stylesheets from the returned object's `stylesheets`
+3. Static `stylesheets` on the component

```

Comment:
I'm wondering what is a use case that developers would use both 2 and 3 here? Is it that technically I can just use the returned object's `stylesheets` so I don't need to modify static `stylesheets` or is there a case that would only work for 3?

Reply:
Static `stylesheets` are the only thing that allows for extending styles from a superclass. If you only use the `render()` API, you would be negating any of the superclass's styles and using only your own.

```js
static stylesheets = [...super.stylesheets, ...myStylesheets]
```

So my thought process is: use `static stylesheets` if you want your subclasses to reuse and extend them. Use `render()` otherwise.